### ISSUE TITLE
Block audit: Button

### ISSUE THREAD

**Issue by sarahmonster** [WordPress/gutenberg#8971]
Block audit

_Note: We'll be doing these audits in waves and editing this as we work through the blocks, so this text will be updated and fleshed out as we progress. [See the full picture here](https://github.com/WordPress/gutenberg/projects/19)._

# Overview

<img width="54" alt="screenshot 2018-08-14 13 42 41" src="https://user-images.githubusercontent.com/376315/44092535-a4781c06-9fc8-11e8-86c8-49e55c916b6e.png">

**Name:** Button
**Description:** Want visitors to click to subscribe, buy, or read more? Get their attention with a button.
***Category:** Layout elements
**CSS class:** `wp-block-button` in frontend and editor
**Can be converted to:** nothing

# States

*Empty:*
<img width="664" alt="screenshot 2018-08-14 13 13 52" src="https://user-images.githubusercontent.com/376315/44092552-ac9c1c02-9fc8-11e8-8a52-21419a24c8d2.png">

*Selected:*
<img width="677" alt="screenshot 2018-08-14 13 40 46" src="https://user-images.githubusercontent.com/376315/44092562-b15b1658-9fc8-11e8-9736-f45734d8cba5.png">

*Unselected:*
<img width="625" alt="screenshot 2018-08-14 13 40 40" src="https://user-images.githubusercontent.com/376315/44092572-b770c740-9fc8-11e8-9781-89704c782a2c.png">

*Placeholder:*
<img width="613" alt="screenshot 2018-08-14 13 43 16" src="https://user-images.githubusercontent.com/376315/44092577-bd075728-9fc8-11e8-98f9-6043544cdfc4.png">

# Primary (toolbar) settings

Block styles

# Secondary (sidebar) settings

Color settings: Background color, text color
Advanced: Additional CSS class

# Frontend appearance

**Gutenberg starter theme:**
<img width="683" alt="screenshot 2018-08-20 11 56 44" src="https://user-images.githubusercontent.com/376315/44336817-8b96ac00-a470-11e8-84de-ef61cac041c8.png">

**Atomic Blocks**
<img width="714" alt="screenshot 2018-08-20 11 55 14" src="https://user-images.githubusercontent.com/376315/44336825-8fc2c980-a470-11e8-95df-325146c2b885.png">

**twentyseventeen:**
<img width="577" alt="screenshot 2018-08-20 11 52 31" src="https://user-images.githubusercontent.com/376315/44336831-93eee700-a470-11e8-8834-8e56644fbe8a.png">

**twentyten:**
<img width="675" alt="screenshot 2018-08-20 11 57 23" src="https://user-images.githubusercontent.com/376315/44336840-994c3180-a470-11e8-94dd-065e579ed0a3.png">

# Bugs/errors

_Individual issues will be opened for these soon_

- [ ] Block styles previews have an awkward shadow effect.

- [ ] Outline ceases to be an outline when a background colour is applied—might make sense to contextually adjust this so the "background" colour is used for the outline. (Seems to be what happens on the front-end.)

- [ ] Suffers from block-level float issues. (#8293)

- [ ] Inconsistent bottom-margin between different alignment settings. (#7812)

- [ ] twentyten's theme styling overrides custom text colour, needs more specific CSS to account for more themes.

# Suggestions

_Individual issues will be opened for these soon_

- [ ] Placeholder on button text is confusing—I expected the text to disappear or highlight. Feels buggy when you can't select to delete it.

- [ ] Arrangement of inputs is rather awkward here. They should at least be spaced out a bit more.

- [ ] Would centre-alignment be a better default here?

- [ ] Should buttons appear on the frontend if no text or link has been added?

- [ ] Hover styling can feel very out of place, depending on the theme styling.

**Comment by webmandesign** [comment]
Hi,

I'm not sure if this is the place I should ask for this, but it's related to Button block.

My question is, would it be possible to set a custom button (link) class? Or add such class? And I mean without using JS, a filter hook in PHP would be preferable.

The thing is that a themes already style buttons and they usually use some generic button class such as `.button` or `.btn` or similar. If we added such class to the block, I think it might simplify styling buttons.

The desired Button block output HTML would then be (using additional `theme-button` class):
```html
<div class="wp-block-button">
	<a class="wp-block-button__link theme-button" href="...">...</a>
</div>
```

Thanks for consideration!

**Comment by webmandesign** [comment]
@0aveRyan Thanks for the info. But the "Additional CSS Class" field doesn't do the trick for me as it doesn't apply the class on the link itself. That way I'm still stuck with custom styling the `.wp-block-button__link` anyway as it can not inherit my `.button` styles.

What I meant was to suggest the button link class filtering functionality. Would be great if theme authors could hook onto the class and change/append to it. Preferably with PHP filter hook, not JS.

**Comment by iamleese** [comment]
I'm not sure if it's been suggested, but an "Open in New Window" option is needed for this block, too.

**Comment by tomusborne** [comment]
Is there a reason the button is rounded by default? I feel like rounding the button this much should be opt-in with a class, instead of the other way around.

**Comment by beckyresler** [comment]
I agree with @iamleese that there needs to be an "Open in New Window" option. It's currently not possible to set a button to do this. If I edit the html to add a **target** on the link, it breaks the button block.

I also agree with @webmandesign that there should be a way to add a class to the link itself rather than the wrapping div. If my theme has button styles for links, I should be able to use them without having to modify the theme.

That brings me to another question ... why is a button wrapped in a div? What if I want to wrap a button with a paragraph?

**Comment by alliecat9lives** [comment]
How about some sort of visual feedback to confirm that I've successfully finished filling out a button? Right now, if I type in my button text, then my URL, then hit Return or click the Apply arrow button, nothing happens to indicate that I've completed the task. 

![no-button-feedback](https://user-images.githubusercontent.com/32844879/50799697-b14b1300-12a2-11e9-969f-3df5120332ce.gif)

As a comparison, when I hyperlink text, there's a visual confirmation that I completed my task.

![hyperlink-feedback](https://user-images.githubusercontent.com/32844879/50799698-b314d680-12a2-11e9-9264-04b316169b98.gif)
